src/datahike/index/persistent_set.cljc

```diff
@@ -1,93 +1,253 @@
+                             dd/cmp-datoms-eavt-quick)
+                           (case index-type
+                             :aevt dd/cmp-temporal-datoms-aevt-quick
+                             :avet dd/cmp-temporal-datoms-avet-quick
+                             dd/cmp-temporal-datoms-eavt-quick))))
+
+(defn index-type->slice-cmp
+  [index-type from to]
+  (let [cmps (->> (case index-type
+                    :eavt [:e :a :v :tx :added]
+                    :aevt [:a :e :v :tx :added]
+                    :avet [:a :v :e :tx :added])
+                  (take-while #(not (and (nil? (% from))
+                                         (nil? (% to)))))
+                  (mapv dd/cmp-val))]
+    (fn [d1 d2]
+      (reduce (fn [old cmp] (let [res (cmp d1 d2)]
+                              (if (nil? res)
+                                (reduced old)
+                                (if (not= 0 res)
+                                  (reduced res)
+                                  old))))
+              0
+              cmps))))
+
+(defn slice [pset from to index-type]
+  (set/slice pset from to (index-type->slice-cmp index-type from to)))
+
+(defn remove-datom [pset datom index-type]
+  (set/disj pset datom (index-type->cmp-quick index-type false)))
+
+(defn insert [pset datom index-type]
+  (if (first (set/slice pset
+                        (dd/datom (.-e datom) (.-a datom) (.-v datom) tx0)
+                        (dd/datom (.-e datom) (.-a datom) (.-v datom) txmax)
+                        (index-type->cmp-quick index-type)))
+    pset
+    (set/conj pset datom (index-type->cmp-quick index-type))))
+
+(defn temporal-insert [pset datom index-type]
+  (set/conj pset datom (index-type->cmp-quick index-type false)))
+
+(defn upsert [pset datom index-type]
+  (-> (if-let [old (first (cond->> (slice pset
+                                          (dd/datom (.-e datom) (.-a datom) nil tx0)
+                                          (dd/datom (.-e datom) (.-a datom) nil txmax)
+                                          index-type)
+                            (= :avet index-type)
+                            (filter #(= (.-e datom) (.-e %)))))]
+        (remove-datom pset old index-type)
+        pset)
+      (set/conj datom (index-type->cmp-quick index-type))))
+
+(defn temporal-upsert [pset datom index-type]
+  (let [{:keys [e a v tx added]} datom
+        old-datoms (cond->> (slice pset
+                                   (dd/datom (.-e datom) (.-a datom) nil tx0)
+                                   (dd/datom (.-e datom) (.-a datom) nil txmax)
+                                   index-type)
+                     (= :avet index-type)
+                     (filter #(= (.-e datom) (.-e %))))
+        {added-datoms true
+         retracted-datoms false} (group-by :added old-datoms)
+        old-val (first (clojure.set/difference (set (map :v added-datoms))
+                                               (map :v retracted-datoms)))]
```

Comment:
This doesn't work if an old value is re-upserted and then replaced, does it? E.g. upsert value `a`, `b`, then `a` again, followed by `c`—if I'm not mistaken, `a` would be added again but not retracted when `c` is upserted.

Reply:
This set logic definitely would not retain order as far as I can see.

Reply:
This had multiple problems, most importantly it is pulling all historical data to do set operations on the full history, which scales in cost with the size of the history for each `[e a]` pair. Instead we should just retrieve the old-value from `eavt` and pass it in from the outside (in general we need to use indices wisely whenever possible). We could not do this with the hitchhiker-tree, because we wanted to avoid lookups and be lazy with writes, but DataScript does the same thing for upserts and the persistent-sorted-set is more than fast enough for that. This change renders upserts in the transaction benchmark sandbox as fast as inserts for me.

Reply:
Note also that this has nothing to do with durability, it also improves the persistent-sorted-set memory speed.